File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        n = len(arr)
        dp = [0] * (n + 1)
        for start in range (n - 1, -1, -1):
            length = 0
            maxNum = -1
            maxSum = 0
            for end in range (start, min(n, start + k)):
                maxNum = max(maxNum, arr[end])
                length += 1
                maxSum = max(maxSum, maxNum * length + dp[end + 1])
            dp[start] = maxSum
        return dp[0]

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        int n = arr.length;
        int[] dp = new int[n + 1];
        for(int start = n - 1; start >= 0; --start) {
            int maxNum = -1, maxSum = 0, length = 0;
            for(int end = start; end < Math.min(n, start + k); ++end) {
                maxNum = Math.max(maxNum, arr[end]);
                length += 1;
                int sum = maxNum * length + dp[end + 1];
                maxSum = Math.max(maxSum, sum);
            }
            dp[start] = maxSum;
        }
        return dp[0];
    }
}